src/main/java/com/vmware/vipclient/i18n/base/instances/LocaleMessage.java

```diff
@@ -71,12 +85,68 @@ public LocaleMessage() {
      */
-    public List<String> getSupportedDisplayNameList(String displayLanguage) {
-        Map<String, String> dispNameMap = new LocaleService().getDisplayNamesFromCLDR(displayLanguage, 
-        		VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+    public List<String> getSupportedDisplayNameList(String displayLocale) {
+        Map<String, String> dispNameMap = getDisplayNamesFromCLDR(displayLocale);
         Collection<String> valueCollection = dispNameMap.values();
         List<String> dispNameList = new ArrayList<String>(valueCollection);
         return dispNameList;
     }
 
+    private Map<String, String> getRegionsFromCLDR(String locale){
+        Map<String, String> regionMap = null;
+        LocaleService localeService = new LocaleService(null);
+        regionMap = localeService.getRegionsFromCLDR(locale);
+        if (regionMap != null) {
+            return regionMap;
+        }
+        if (!LocaleUtility.isDefaultLocale(locale)) {
+            logger.info("Can't find regions for locale [{}], look for English regions as fallback!", locale);
+            regionMap = localeService.getRegionsFromCLDR(LocaleUtility.getDefaultLocale().toLanguageTag());
+            if (regionMap != null) {
+                new FormattingCacheService().addRegions(locale, regionMap);
+                logger.debug("Default locale's regions is cached for locale [{}]!\n\n", locale);
+            }
+        }
+        return regionMap;
+    }
+
+    private Map<String, String> getDisplayNamesFromCLDR(String locale) {
+        Map<String, String> dispMap = new HashMap<String, String>();
+        if(locale == null || locale.isEmpty()) {
+            logger.warn("Locale is empty!");
+            return dispMap;
+        }
+        BaseDTO dto = new BaseDTO();
+        if(cfg != null) {
+            dto.setProductID(cfg.getProductName());
+            dto.setVersion(cfg.getVersion());
+        }else{
+            dto.setProductID(VIPCfg.getInstance().getProductName());
+            dto.setVersion(VIPCfg.getInstance().getVersion());
+        }
+        LocaleService localeService = new LocaleService(dto);
+        dispMap = localeService.getSupportedDisplayNames(locale);
+        if(dispMap != null && !dispMap.isEmpty()){
+            return dispMap;
+        }
+        if (!LocaleUtility.isDefaultLocale(locale)) {
+            logger.info("Can't find supported languages for locale [{}], look for English languages as fallback!", locale);
+            Locale fallbackLocale = LocaleUtility.getDefaultLocale();
+            dispMap = localeService.getSupportedDisplayNames(fallbackLocale.toLanguageTag());
+            if (dispMap != null && dispMap.size() > 0) {
+                new FormattingCacheService().addSupportedLanguages(dto, locale, dispMap);
+                logger.debug("Default locale's displayNames is cached for product [{}], version [{}], locale [{}]!\n\n",
+                        dto.getProductID(), dto.getVersion(), locale);
```

Comment:
This is changed according to Jessiely's suggestion, cache default locale's content as requested locale, or will always try to fetch the requested locale first before using the fallback locale that is already in the cache, which may increase product UI loading time.


Below is Jessiely's original words:

Put this fallback logic in LocaleMessage so that this method here in LocaleService is always for the requested language only.

Once you move the fallback logic into LocaleMessage, update the cache in there in such a way that the key with the requested language is mapped to the FormatCacheItem of the fallback language.
c.put(<cacheKeyOfRequestedLanguage>, new FormatCacheItem(<displayMapofFallbackLanguage>));

Without the above proposed changes, your code will always try to fetch the requested locale first before using the fallback locale that is already in the cache. With the above proposed changes, any succeeding calls for the requested language will immediately return the cached fallback locale until it expires.

Note: You can either put the above logic in LocaleMessage as I propose (just like how it is in TranslationMessage.getMessages), or create another method here in LocaleService to keep the logic in the "service layer".